File: migracao-go/internal/application/programacao/frentes.go
package programacao

import (
	"context"
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"siqueiracampos/servidor/internal/domain/identidade"
	dominio "siqueiracampos/servidor/internal/domain/programacao"
)

var reCorHex = regexp.MustCompile(`^#[0-9a-fA-F]{6}$`)

type EntradaFrente struct {
	Nome, Cor, Logo, ObraCodigo, Colunas string
}

func validarFrente(e EntradaFrente) (nome, cor string, colunas int, err error) {
	nome = strings.TrimSpace(e.Nome)
	if len(nome) < 2 {
		return "", "", 0, fmt.Errorf("informe o nome do cliente ou da frente")
	}
	cor = strings.TrimSpace(e.Cor)
	if !reCorHex.MatchString(cor) {
		return "", "", 0, fmt.Errorf("escolha uma cor válida")
	}
	colunas = 1
	if strings.TrimSpace(e.Colunas) != "" {
		colunas, err = strconv.Atoi(strings.TrimSpace(e.Colunas))
		if err != nil || colunas < 1 || colunas > 4 {
			return "", "", 0, fmt.Errorf("colunas deve ser entre 1 e 4")
		}
	}
	return nome, cor, colunas, nil
}

// CriarFrente não reduz a logo enviada (o app original usa `sharp` para isso — sem
// equivalente direto em Go nesta fatia); a tela deve limitar o tamanho antes de enviar.
func (g *Gerenciador) CriarFrente(ctx context.Context, s identidade.Sessao, e EntradaFrente) (*dominio.Frente, error) {
	if !identidade.PodeLancar(s.Cargo) {
		return nil, fmt.Errorf("seu cargo não cadastra frentes")
	}
	nome, cor, colunas, err := validarFrente(e)
	if err != nil {
		return nil, err
	}
	todas, err := g.Repo.ListarFrentes(ctx, false)
	if err != nil {
		return nil, err
	}
	ordem := 0
	for _, f := range todas {
		if f.Ordem > ordem {
			ordem = f.Ordem
		}
	}
	f := &dominio.Frente{
		Nome: nome, Cor: cor, Logo: strPtr(e.Logo), ObraCodigo: strPtr(e.ObraCodigo),
		Colunas: colunas, Ordem: ordem + 1, Ativa: true,
	}
	if err = g.Repo.CriarFrente(ctx, f); err != nil {
		return nil, erroFrenteDuplicada(err, nome)
	}
	return f, nil
}

func (g *Gerenciador) EditarFrente(ctx context.Context, s identidade.Sessao, id string, e EntradaFrente) error {
	if !identidade.PodeLancar(s.Cargo) {
		return fmt.Errorf("seu cargo não edita frentes")
	}
	atual, err := g.Repo.BuscarFrente(ctx, id)
	if err != nil {
		return err
	}
	if atual == nil {
		return fmt.Errorf("frente não encontrada")
	}
	nome, cor, colunas, err := validarFrente(e)
	if err != nil {
		return err
	}
	atual.Nome, atual.Cor, atual.Colunas = nome, cor, colunas
	atual.Logo, atual.ObraCodigo = strPtr(e.Logo), strPtr(e.ObraCodigo)
	if err = g.Repo.AtualizarFrente(ctx, atual); err != nil {
		return erroFrenteDuplicada(err, nome)
	}
	return nil
}

func erroFrenteDuplicada(err error, nome string) error {
	if strings.Contains(strings.ToLower(err.Error()), "unique") {
		return fmt.Errorf("já existe uma frente chamada %q", nome)
	}
	return err
}

// AlternarFrente desativa em vez de apagar quando a frente já apareceu em algum dia —
// apagar levaria junto as escalas daquele dia, e a programação de uma terça passada
// deixaria de dizer quem estava onde.
func (g *Gerenciador) AlternarFrente(ctx context.Context, s identidade.Sessao, id string, ativa bool) error {
	if !identidade.PodeLancar(s.Cargo) {
		return fmt.Errorf("seu cargo não altera frentes")
	}
	return g.Repo.AlternarFrente(ctx, id, ativa)
}

// ApagarFrente só serve para a frente criada por engano, que nunca entrou num dia.
func (g *Gerenciador) ApagarFrente(ctx context.Context, s identidade.Sessao, id string) error {
	if !identidade.PodeLancar(s.Cargo) {
		return fmt.Errorf("seu cargo não apaga frentes")
	}
	usos, err := g.Repo.ContarUsosFrente(ctx, id)
	if err != nil {
		return err
	}
	if usos > 0 {
		return fmt.Errorf("esta frente já foi usada em %d lançamentos; desative em vez de apagar", usos)
	}
	return g.Repo.ApagarFrente(ctx, id)
}

// MoverFrente sobe ou desce a frente — é a ordem em que as colunas saem no quadro/imagem.
func (g *Gerenciador) MoverFrente(ctx context.Context, s identidade.Sessao, id, direcao string) error {
	if !identidade.PodeLancar(s.Cargo) {
		return fmt.Errorf("seu cargo não reordena frentes")
	}
	todas, err := g.Repo.ListarFrentes(ctx, false)
	if err != nil {
		return err
	}
	i := -1
	for idx, f := range todas {
		if f.ID == id {
			i = idx
			break
		}
	}
	if i < 0 {
		return fmt.Errorf("frente não encontrada")
	}
	j := i - 1
	if direcao == "baixo" {
		j = i + 1
	}
	if j < 0 || j >= len(todas) {
		return nil
	}
	return g.Repo.ReordenarFrentes(ctx, todas[i].ID, todas[j].Ordem, todas[j].ID, todas[i].Ordem)
}

File: apps/programacao/src/actions/frentes.ts
'use server'

import { z } from 'zod'
import { prisma } from '@/lib/prisma'
import { exigirLancamento } from '@/lib/auth'
import { revalidarTelas } from '@/lib/revalidar'

export type Resultado<T = void> = { ok: true; dados: T } | { ok: false; erro: string }

/**
 * Campo opcional que aceita ausente, vazio E null.
 *
 * O `null` importa: a tela manda `campo || null` para dizer "não preencheu", e um esquema
 * que só aceita string recusa isso com "expected string, received null" — mensagem que não
 * diz nada a quem está preenchendo o formulário.
 */
const opcional = z.string().trim().nullish().transform((v) => (v ? v : null))

const esquema = z.object({
  nome: z.string().trim().min(2, 'Informe o nome do cliente ou da frente.'),
  cor: z.string().trim().regex(/^#[0-9a-fA-F]{6}$/, 'Escolha uma cor.'),
  logo: opcional,
  colunas: z.coerce.number().int().min(1).max(4).default(1),
  obraCodigo: opcional,
})

async function otimizarLogo(logo: string | null): Promise<string | null> {
  if (!logo) return null
  const separador = logo.indexOf(',')
  if (!logo.startsWith('data:') || separador < 0) throw new Error('A logo precisa ser uma imagem válida.')
  const bruto = Buffer.from(logo.slice(separador + 1), 'base64')
  if (bruto.length > 12 * 1024 * 1024) throw new Error('A logo deve ter no máximo 12 MB.')
  const { default: sharp } = await import('sharp')
  const webp = await sharp(bruto, { animated: false })
    .resize({ width: 640, height: 360, fit: 'inside', withoutEnlargement: true })
    .webp({ quality: 88 })
    .toBuffer()
  return `data:image/webp;base64,${webp.toString('base64')}`
}

function revalidar() {
  revalidarTelas('/frentes', '/')
}

export async function criarFrente(entrada: unknown): Promise<Resultado<{ id: string }>> {
  await exigirLancamento()

  const parsed = esquema.safeParse(entrada)
  if (!parsed.success) return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }

  try {
    const ultima = await prisma.frente.findFirst({ orderBy: { ordem: 'desc' }, select: { ordem: true } })
    const logo = await otimizarLogo(parsed.data.logo)
    const criada = await prisma.frente.create({
      data: { ...parsed.data, logo, ordem: (ultima?.ordem ?? 0) + 1 },
    })
    revalidar()
    return { ok: true, dados: { id: criada.id } }
  } catch (e) {
    const msg = e instanceof Error && e.message.includes('Unique')
      ? `Já existe uma frente chamada "${parsed.data.nome}".`
      : 'Falha ao criar a frente.'
    return { ok: false, erro: msg }
  }
}

export async function editarFrente(id: string, entrada: unknown): Promise<Resultado> {
  await exigirLancamento()

  const parsed = esquema.safeParse(entrada)
  if (!parsed.success) return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }

  try {
    const logo = await otimizarLogo(parsed.data.logo)
    await prisma.frente.update({ where: { id }, data: { ...parsed.data, logo } })
    revalidar()
    return { ok: true, dados: undefined }
  } catch (e) {
    const msg = e instanceof Error && e.message.includes('Unique')
      ? `Já existe uma frente chamada "${parsed.data.nome}".`
      : 'Falha ao salvar a frente.'
    return { ok: false, erro: msg }
  }
}

/**
 * Desativa em vez de apagar quando a frente já apareceu em algum dia.
 *
 * Apagar levaria junto as escalas daquele dia — e a programação de uma terça passada
 * deixaria de dizer quem estava onde. Frente desativada some do quadro de amanhã e continua
 * nos dias antigos, que é exatamente o comportamento que se espera de um cliente que
 * terminou o contrato.
 */
export async function alternarFrente(id: string, ativa: boolean): Promise<Resultado> {
  await exigirLancamento()
  try {
    await prisma.frente.update({ where: { id }, data: { ativa } })
    revalidar()
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao alterar a frente.' }
  }
}

/** Apaga de vez — só serve para a frente criada por engano, que nunca entrou num dia. */
export async function apagarFrente(id: string): Promise<Resultado> {
  await exigirLancamento()

  try {
    const usos = await prisma.escala.count({ where: { frenteId: id } })
    const recursos = await prisma.recurso.count({ where: { frenteId: id } })
    if (usos + recursos > 0) {
      return {
        ok: false,
        erro:
          `Esta frente já foi usada em ${usos + recursos} lançamentos. ` +
          'Desative em vez de apagar — apagar levaria junto a programação dos dias em que ela apareceu.',
      }
    }

    await prisma.frente.delete({ where: { id } })
    revalidar()
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao apagar a frente.' }
  }
}

/** Sobe ou desce a frente — é a ordem em que as colunas saem na imagem. */
export async function moverFrente(id: string, direcao: 'cima' | 'baixo'): Promise<Resultado> {
  await exigirLancamento()

  try {
    const todas = await prisma.frente.findMany({ orderBy: { ordem: 'asc' } })
    const i = todas.findIndex((f) => f.id === id)
    if (i < 0) return { ok: false, erro: 'Frente não encontrada.' }

    const j = direcao === 'cima' ? i - 1 : i + 1
    if (j < 0 || j >= todas.length) return { ok: true, dados: undefined }

    // Troca as posições numa transação: sem ela, uma falha no meio deixaria duas frentes
    // com a mesma ordem e a imagem sairia com as colunas em ordem imprevisível.
    await prisma.$transaction([
      prisma.frente.update({ where: { id: todas[i].id }, data: { ordem: todas[j].ordem } }),
      prisma.frente.update({ where: { id: todas[j].id }, data: { ordem: todas[i].ordem } }),
    ])
    revalidar()
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao reordenar.' }
  }
}
